### PR TITLE
add instrumentation probes for easier performance analysis

### PR DESCRIPTION
**Pull request by tomerd**
motivation: help with performance analsysis

changes:
* define a few more workspace delegates
* in debug mode, emit additional performance information
* add signpost helpers

### CODE REVIEW

File: Sources/Commands/PackageTools/Update.swift [hunk 1]
```diff
@@ -41,16 +42,18 @@ extension SwiftPackageTool {
                 observabilityScope: swiftTool.observabilityScope
             )
             
-            // try to load the graph which will emit any errors
-            if !swiftTool.observabilityScope.errorsReported {
-                _ = try workspace.loadPackageGraph(
-                    rootInput: swiftTool.getWorkspaceRoot(),
-                    observabilityScope: swiftTool.observabilityScope
-                )
+            do {
```
Comment (MaxDesiatov) [inline Sources/Commands/PackageTools/Update.swift:45]: Does this `do` block have any effect without `catch`?

**Review by neonichu — APPROVED** [review]

File: Sources/Commands/PackageTools/Update.swift (hunk 1 above)
Comment (tomerd) [reply Sources/Commands/PackageTools/Update.swift:45]: its to scope the `start` variable

Comment (MaxDesiatov) [reply Sources/Commands/PackageTools/Update.swift:45]: Sorry, I don't see `start` declared in this block, but I see it declared in other functions. Could that be missing here by any chance, or am I misunderstanding how `do` can be applied?

Comment (tomerd) [reply Sources/Commands/PackageTools/Update.swift:45]: oops you are right, I removed that code, so that is left over

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci test windows

**Pull request merged**